File: src/open_workshop_storage/api/routes/files.py
from __future__ import annotations

import asyncio
import os
from functools import lru_cache
from typing import Callable, Optional
from urllib.parse import urlparse

import anyio
import ow_config as config
from fastapi import APIRouter, File, Form, Request, UploadFile
from fastapi.responses import FileResponse, PlainTextResponse
from pydantic import BaseModel, Field

from ...core.blurhash_service import get_or_compute_blurhash_for_key
from ...core.context import ServiceContext
from ...services import access_client
from ...service_factory import get_current_service_context
from ...utils import image_file_to_blurhash

router = APIRouter()
_context_provider: Optional[Callable[[], ServiceContext]] = None


def configure_context_provider(provider: Callable[[], ServiceContext]) -> None:
    global _context_provider
    _context_provider = provider


def _ctx() -> ServiceContext:
    try:
        return get_current_service_context()
    except RuntimeError:
        if _context_provider is None:
            raise RuntimeError("file context provider is not configured")
        return _context_provider()


def _client_host(request: Request) -> str:
    return request.client.host if request.client else "unknown"


def _error_response(status_code: int, content: str) -> PlainTextResponse:
    return PlainTextResponse(status_code=status_code, content=content)


def _read_int_setting(name: str, default: int) -> int:
    raw_value = getattr(config, name, default)
    try:
        return max(0, int(raw_value))
    except (TypeError, ValueError):
        return max(0, int(default))


# Keep the in-process cache modest: it is replicated in every worker process,
# so a very large default can inflate RSS quickly on busy distributors.
BLURHASH_CACHE_SIZE = _read_int_setting("BLURHASH_CACHE_SIZE", 4096)


class BlurhashBatchRequest(BaseModel):
    paths: list[str] = Field(default_factory=list, description="Storage download URLs or storage-relative paths.")


class BlurhashItemRead(BaseModel):
    path: str
    blurhash: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None


class BlurhashBatchResponse(BaseModel):
    items: list[BlurhashItemRead] = Field(default_factory=list)


def _normalize_blurhash_target(raw_path: str) -> Optional[tuple[str, str]]:
    value = str(raw_path or "").strip()
    if not value:
        return None

    parsed = urlparse(value)
    if parsed.scheme and parsed.netloc:
        value = parsed.path

    value = value.lstrip("/")
    if not value.startswith("download/"):
        return None

    parts = value.split("/", 2)
    if len(parts) < 3:
        return None

    storage_type = parts[1]
    storage_path = parts[2]
    if not storage_type or not storage_path:
        return None
    return storage_type, storage_path


@lru_cache(maxsize=BLURHASH_CACHE_SIZE)
def _blurhash_for_file(
    real_path: str,
    mtime_ns: int,
    size: int,
    components_x: int = 4,
    components_y: int = 3,
    max_dimension: int = 64,
) -> tuple[str, int, int]:
    del mtime_ns, size
    return image_file_to_blurhash(
        real_path,
        components_x=components_x,
        components_y=components_y,
        max_dimension=max_dimension,
    )


def _prepare_blurhash_item(
    ctx: ServiceContext,
    raw_path: str,
) -> tuple[BlurhashItemRead, Optional[tuple[str, int, int]]]:
    item = BlurhashItemRead(path=raw_path)
    normalized = _normalize_blurhash_target(raw_path)
    if normalized is None:
        return item, None

    storage_type, storage_path = normalized
    if not ctx.tools.is_allowed_type(storage_type):
        return item, None

    base_dir = os.path.join(ctx.main_dir, storage_type)
    try:
        real_path = ctx.tools.safe_path(base_dir, storage_path)
    except ValueError:
        return item, None

    if not os.path.isfile(real_path):
        return item, None

    try:
        stat_result = os.stat(real_path)
    except (OSError, ValueError):
        return item, None

    return item, (real_path, stat_result.st_mtime_ns, stat_result.st_size)


async def _compute_blurhash_for_key(
    ctx: ServiceContext,
    key: tuple[str, int, int],
) -> Optional[tuple[str, int, int]]:
    return await get_or_compute_blurhash_for_key(ctx, key, _blurhash_for_file)


@router.post(
    "/blurhashes",
    tags=["Files"],
    summary="Generate blurhashes for stored images",
    description=(
        "Generates BlurHash strings for a batch of stored image URLs or storage-relative download paths.\n\n"
        "Use full storage download URLs from the website or relative paths like `download/resource/...`."
    ),
    status_code=200,
    response_model=BlurhashBatchResponse,
    response_description="Batch of BlurHash results.",
    responses={
        200: {
            "description": "BlurHash batch generated successfully",
        },
    },
)
async def blurhashes(request: Request, payload: BlurhashBatchRequest) -> BlurhashBatchResponse:
    ctx = _ctx()
    client = _client_host(request)
    ctx.logger.info("blurhash batch request items=%s client=%s", len(payload.paths), client)
    prepared_items = [_prepare_blurhash_item(ctx, raw_path) for raw_path in payload.paths]
    unique_keys = list(dict.fromkeys(key for _, key in prepared_items if key is not None))
    blurhash_results: dict[tuple[str, int, int], Optional[tuple[str, int, int]]] = {}

    if unique_keys:
        computed = await asyncio.gather(*(_compute_blurhash_for_key(ctx, key) for key in unique_keys))
        blurhash_results = {key: result for key, result in zip(unique_keys, computed)}

    items: list[BlurhashItemRead] = []
    for item, key in prepared_items:
        if key is not None:
            blurhash_data = blurhash_results.get(key)
            if blurhash_data is not None:
                blurhash, width, height = blurhash_data
                item.blurhash = blurhash
                item.width = width
                item.height = height
        items.append(item)

    return BlurhashBatchResponse(items=items)


@router.api_route(
    "/download/{type}/{path:path}",
    methods=["GET", "HEAD"],
    tags=["Files"],
    summary="Download stored file",
    description=(
        "Downloads a stored file. For archive/mod downloads access is validated via the access service. "
        "Optional query param `filename` can be used to override download name (safe chars only)."
    ),
    status_code=200,
    response_class=FileResponse,
    responses={
        200: {
            "description": "File send successfully",
            "content": {"application/octet-stream": {}},
        },
        400: {
            "description": "Invalid type",
            "content": {"text/plain": {"example": "Invalid type"}},
        },
        403: {
            "description": "Access denied",
            "content": {"text/plain": {"example": "Access denied"}},
        },
        423: {
            "description": "Access denied",
            "content": {"text/plain": {"example": "Access denied"}},
        },
        404: {
            "description": "File not found on server",
            "content": {"text/plain": {"example": "File not found"}},
        },
        503: {
            "description": "Access service unavailable",
            "content": {"text/plain": {"example": "Access service unavailable"}},
        },
    },
)
async def download(request: Request, type: str, path: str, filename: Optional[str] = None):
    ctx = _ctx()
    client = request.client.host if request.client else "unknown"
    ctx.logger.info("download request type=%s path=%s client=%s", type, path, client)
    if not ctx.tools.is_allowed_type(type):
        ctx.logger.warning("download invalid type=%s path=%s client=%s", type, path, client)
        return PlainTextResponse(status_code=400, content="Invalid type")
    base_dir = os.path.join(ctx.main_dir, type)
    try:
        real_path = ctx.tools.safe_path(base_dir, path)
    except ValueError:
        ctx.logger.warning("download path traversal type=%s path=%s client=%s", type, path, client)
        return PlainTextResponse(status_code=423, content="Access denied")

    if not os.path.isfile(real_path):
        ctx.logger.info("download not found type=%s path=%s client=%s", type, path, client)
        return PlainTextResponse(status_code=404, content="File not found")

    download_name = ctx.tools.build_download_filename(filename, real_path)

    async def file_response_with_meta() -> FileResponse:
        response = FileResponse(real_path, filename=download_name)
        if request.method == "HEAD" and type == "archive" and path.startswith("mods/"):
            try:
                _, is_encrypted, archive_entries = await anyio.to_thread.run_sync(
                    ctx.tools.probe_archive,
                    real_path,
                )
                if not is_encrypted:
                    unpacked_bytes = await anyio.to_thread.run_sync(
                        ctx.tools.archive_entries_unpacked_bytes,
                        archive_entries,
                    )
                    if unpacked_bytes is not None:
                        response.headers["X-Unpacked-Bytes"] = str(unpacked_bytes)
            except Exception:
                ctx.logger.warning(
                    "download unpacked header failed type=%s path=%s client=%s",
                    type,
                    path,
                    client,
                )
        return response

    if type == "archive" and path.startswith("mods/"):
        parts = path.split("/", 2)
        if len(parts) < 2:
            ctx.logger.info("download not found (bad mod id) type=%s path=%s client=%s", type, path, client)
            return _error_response(404, "File not found")
        try:
            mod_id = int(parts[1])
        except ValueError:
            ctx.logger.info("download not found (bad mod id) type=%s path=%s client=%s", type, path, client)
            return _error_response(404, "File not found")
        try:
            access_result = await access_client.resolve_mod_download_access(
                request=request,
                mod_id=mod_id,
                access_service_url=ctx.config.ACCESS_SERVICE_URL,
                timeout_seconds=int(ctx.config.ACCESS_SERVICE_TIMEOUT_SECONDS),
            )
            if access_result.allowed:
                ctx.logger.info(
                    "download allowed mod_id=%s type=%s path=%s client=%s",
                    mod_id,
                    type,
                    path,
                    client,
                )
                return await file_response_with_meta()
            ctx.logger.warning(
                "download denied mod_id=%s type=%s path=%s client=%s reason_code=%s",
                mod_id,
                type,
                path,
                client,
                access_result.reason_code,
            )
            return _error_response(403, access_result.reason or "Access denied")
        except access_client.AccessServiceError as exc:
            status_code = exc.status_code if exc.status_code is not None else 503
            ctx.logger.warning(
                "access service error status=%s mod_id=%s type=%s path=%s client=%s message=%s",
                status_code,
                mod_id,
                type,
                path,
                client,
                str(exc),
            )
            return _error_response(status_code, str(exc))
        except (asyncio.TimeoutError, TypeError, ValueError):
            ctx.logger.warning(
                "access service unavailable network error mod_id=%s type=%s path=%s client=%s",
                mod_id,
                type,
                path,
                client,
            )
            return _error_response(503, "Access service unavailable")

    ctx.logger.info("download ok type=%s path=%s client=%s", type, path, client)
    return await file_response_with_meta()


@router.post(
    "/upload",
    tags=["Files"],
    summary="Upload file to Storage (internal)",
    description=(
        "Internal upload endpoint for Manager. "
        "Accepts multipart form-data with file, type, path and file_kind. "
        "Requires upload token."
    ),
    status_code=201,
    response_class=PlainTextResponse,
    response_model=str,
    responses={
        201: {
            "description": "File uploaded successfully",
            "content": {"text/plain": {"example": "file/is/saved/as.tmp"}},
            "model": str,
        },
        401: {
            "description": "Token not found",
            "content": {"text/plain": {"example": "Token not found"}},
            "model": str,
        },
        400: {
            "description": "Invalid type",
            "content": {"text/plain": {"example": "Invalid type"}},
            "model": str,
        },
    },
)
async def upload(
    request: Request,
    file: UploadFile = File(),
    storage_type: str = Form(alias="type"),
    storage_path: str = Form(alias="path"),
    file_kind: str = Form("bin"),
    token: str = Form(),
):
    ctx = _ctx()
    client = _client_host(request)
    ctx.logger.info(
        "upload request type=%s path=%s file_kind=%s filename=%s client=%s",
        storage_type,
        storage_path,
        file_kind,
        file.filename,
        client,
    )
    if not token:
        ctx.logger.warning(
            "upload denied (token missing) type=%s path=%s client=%s",
            storage_type,
            storage_path,
            client,
        )
        return _error_response(401, "Token not found")
    if not await anyio.to_thread.run_sync(ctx.tools.check_token, "upload_file", token):
        ctx.logger.warning("upload denied (token) type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(403, "Access denied")
    if not ctx.tools.is_allowed_upload_type(storage_type):
        ctx.logger.warning("upload invalid type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(400, "Invalid type")
    normalized_file_kind = ctx.tools.normalize_file_kind(file_kind, default="")
    if not normalized_file_kind:
        ctx.logger.warning(
            "upload invalid file_kind=%s type=%s path=%s client=%s",
            file_kind,
            storage_type,
            storage_path,
            client,
        )
        return _error_response(400, "Invalid file kind")
    if storage_type == "avatar" and normalized_file_kind != "img":
        return _error_response(400, "Avatar requires image file kind")

    base_dir = os.path.join(ctx.main_dir, storage_type)
    try:
        real_path = ctx.tools.safe_path(base_dir, storage_path)
    except ValueError:
        ctx.logger.warning("upload path traversal type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(423, "Access denied")
    os.makedirs(os.path.dirname(real_path), exist_ok=True)

    if normalized_file_kind == "img":
        if not storage_path.lower().endswith(".webp"):
            return _error_response(400, "Image storage path must end with .webp")
        raw_bytes = await file.read()
        try:
            webp_bytes = await anyio.to_thread.run_sync(ctx.tools.image_bytes_to_webp, raw_bytes)
        except ValueError:
            return _error_response(400, "Image expected")

        def _write_bytes_sync() -> None:
            with open(real_path, "wb") as out_file:
                out_file.write(webp_bytes)

        await anyio.to_thread.run_sync(_write_bytes_sync)
    else:
        await anyio.to_thread.run_sync(ctx.tools.copy_fileobj_to_path, file.file, real_path)
    ctx.logger.info(
        "upload saved type=%s path=%s file_kind=%s client=%s",
        storage_type,
        storage_path,
        normalized_file_kind,
        client,
    )
    return storage_path


@router.delete(
    "/delete",
    tags=["Files"],
    summary="Delete file from Storage (internal)",
    description=(
        "Internal delete endpoint for Manager. " "Deletes file and empty parent folders. Requires delete token."
    ),
    status_code=200,
    response_class=PlainTextResponse,
    response_model=str,
    responses={
        200: {
            "description": "File deleted successfully",
            "content": {"text/plain": {"example": "File deleted"}},
            "model": str,
        },
        401: {
            "description": "Token not found",
            "content": {"text/plain": {"example": "Token not found"}},
            "model": str,
        },
        400: {
            "description": "Invalid type",
            "content": {"text/plain": {"example": "Invalid type"}},
            "model": str,
        },
        404: {
            "description": "File not found on server",
            "content": {"text/plain": {"example": "File not found"}},
            "model": str,
        },
    },
)
async def delete(
    request: Request,
    storage_type: str = Form(alias="type"),
    storage_path: str = Form(alias="path"),
    token: str = Form(),
):
    ctx = _ctx()
    client = _client_host(request)
    ctx.logger.info("delete request type=%s path=%s client=%s", storage_type, storage_path, client)
    if not token:
        ctx.logger.warning(
            "delete denied (token missing) type=%s path=%s client=%s",
            storage_type,
            storage_path,
            client,
        )
        return _error_response(401, "Token not found")
    if not await anyio.to_thread.run_sync(ctx.tools.check_token, "delete_file", token):
        ctx.logger.warning("delete denied (token) type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(403, "Access denied")
    if not ctx.tools.is_allowed_type(storage_type):
        ctx.logger.warning("delete invalid type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(400, "Invalid type")

    base_dir = os.path.join(ctx.main_dir, storage_type)
    try:
        real_path = ctx.tools.safe_path(base_dir, storage_path)
    except ValueError:
        ctx.logger.warning("delete path traversal type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(403, "Access denied")

    def delete_file_and_parent_folders(file_path: str, root_dir: str):
        if not os.path.isfile(file_path):
            ctx.logger.info("delete not found type=%s path=%s client=%s", storage_type, storage_path, client)
            return _error_response(404, "File not found")
        os.remove(file_path)
        folder_path = os.path.dirname(file_path)
        root_dir = os.path.abspath(root_dir)
        while folder_path and os.path.commonpath([folder_path, root_dir]) == root_dir and folder_path != root_dir:
            if not os.listdir(folder_path):
                os.rmdir(folder_path)
                folder_path = os.path.dirname(folder_path)
            else:
                break

        ctx.logger.info("delete ok type=%s path=%s client=%s", storage_type, storage_path, client)
        return _error_response(200, "File deleted")

    return await anyio.to_thread.run_sync(delete_file_and_parent_folders, real_path, base_dir)

File: src/handlers/files.rs
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::sync::Arc;

use axum::body::Body;
use axum::extract::{Extension, Multipart, Path as AxumPath, Query};
use axum::http::header::{CONTENT_DISPOSITION, CONTENT_LENGTH, CONTENT_TYPE};
use axum::http::{HeaderMap, Method, StatusCode, Uri};
use axum::response::{IntoResponse, Response};
use axum::Json;
use futures_util::future::join_all;
use serde::{Deserialize, Serialize};
use tokio::fs;
use tokio_util::io::ReaderStream;

use crate::access_client;
use crate::archive::{archive_entries_unpacked_bytes, probe_archive};
use crate::auth;
use crate::blurhash_service::get_or_compute_blurhash_for_key;
use crate::fs_utils::{
    build_download_filename, is_allowed_type, is_allowed_upload_type, normalize_file_kind,
    safe_path,
};
use crate::images::image_bytes_to_webp;
use crate::runtime::AppState;

#[derive(Debug, Clone, Deserialize)]
pub struct BlurhashBatchRequest {
    pub paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlurhashItemRead {
    pub path: String,
    pub blurhash: Option<String>,
    pub width: Option<u32>,
    pub height: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlurhashBatchResponse {
    pub items: Vec<BlurhashItemRead>,
}

#[derive(Debug, Deserialize)]
pub struct DownloadPath {
    storage_type: String,
    path: String,
}

#[derive(Debug, Deserialize)]
pub struct DownloadQuery {
    filename: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct DeleteRequest {
    #[serde(rename = "type")]
    storage_type: String,
    path: String,
    token: Option<String>,
}

fn error_response(status: StatusCode, content: impl Into<String>) -> Response {
    (status, content.into()).into_response()
}

fn normalize_blurhash_target(raw_path: &str) -> Option<(String, String)> {
    let mut value = raw_path.trim().to_string();
    if value.is_empty() {
        return None;
    }

    if let Ok(uri) = value.parse::<Uri>() {
        if uri.scheme_str().is_some() && uri.authority().is_some() {
            value = uri.path().to_string();
        }
    }

    let value = value.trim_start_matches('/').to_string();
    if !value.starts_with("download/") {
        return None;
    }
    let mut parts = value.splitn(3, '/');
    let _ = parts.next();
    let storage_type = parts.next()?.to_string();
    let storage_path = parts.next()?.to_string();
    if storage_type.is_empty() || storage_path.is_empty() {
        return None;
    }
    Some((storage_type, storage_path))
}

async fn prepare_blurhash_item(
    state: Arc<AppState>,
    raw_path: String,
) -> (BlurhashItemRead, Option<(PathBuf, u128, u64)>) {
    let item = BlurhashItemRead {
        path: raw_path.clone(),
        blurhash: None,
        width: None,
        height: None,
    };
    let Some((storage_type, storage_path)) = normalize_blurhash_target(&raw_path) else {
        return (item, None);
    };
    if !is_allowed_type(&storage_type) {
        return (item, None);
    }

    let base_dir = state.main_dir.join(&storage_type);
    let Ok(real_path) = safe_path(&base_dir, &storage_path) else {
        return (item, None);
    };
    if !real_path.is_file() {
        return (item, None);
    }
    let Ok(stat_result) = std::fs::metadata(&real_path) else {
        return (item, None);
    };
    let key = (
        real_path,
        stat_result
            .modified()
            .ok()
            .and_then(|value| value.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|value| value.as_nanos())
            .unwrap_or(0),
        stat_result.len(),
    );
    (item, Some(key))
}

pub async fn blurhashes(
    Extension(state): Extension<Arc<AppState>>,
    Json(payload): Json<BlurhashBatchRequest>,
) -> impl IntoResponse {
    let prepared = join_all(
        payload
            .paths
            .into_iter()
            .map(|path| prepare_blurhash_item(state.clone(), path)),
    )
    .await;
    let mut seen = HashSet::new();
    let mut unique_keys = Vec::new();
    for (_, key) in &prepared {
        if let Some(key) = key.clone() {
            if seen.insert(key.clone()) {
                unique_keys.push(key);
            }
        }
    }

    let computed = join_all(
        unique_keys
            .iter()
            .cloned()
            .map(|key| get_or_compute_blurhash_for_key(state.clone(), key, 4, 3, 64)),
    )
    .await;

    let blurhash_results: HashMap<_, _> =
        unique_keys.into_iter().zip(computed.into_iter()).collect();

    let mut items = Vec::new();
    for (mut item, key) in prepared {
        if let Some(key) = key {
            if let Some(Some((blurhash, width, height))) = blurhash_results.get(&key).cloned() {
                item.blurhash = Some(blurhash);
                item.width = Some(width);
                item.height = Some(height);
            }
        }
        items.push(item);
    }

    Json(BlurhashBatchResponse { items })
}

async fn open_file_response(
    real_path: PathBuf,
    download_name: String,
    head_only: bool,
    unpacked_bytes: Option<u64>,
) -> Result<Response, Response> {
    let file_size = fs::metadata(&real_path)
        .await
        .map_err(|_| error_response(StatusCode::NOT_FOUND, "File not found"))?
        .len();
    let mut builder = Response::builder()
        .status(StatusCode::OK)
        .header(CONTENT_TYPE, "application/octet-stream")
        .header(
            CONTENT_DISPOSITION,
            format!(
                "attachment; filename=\"{}\"",
                download_name.replace('"', "")
            ),
        );
    builder = builder.header(CONTENT_LENGTH, file_size.to_string());
    if let Some(unpacked_bytes) = unpacked_bytes {
        builder = builder.header("X-Unpacked-Bytes", unpacked_bytes.to_string());
    }
    if head_only {
        return builder
            .body(Body::empty())
            .map_err(|err| error_response(StatusCode::INTERNAL_SERVER_ERROR, err.to_string()));
    }
    let file = fs::File::open(&real_path)
        .await
        .map_err(|_| error_response(StatusCode::NOT_FOUND, "File not found"))?;
    let body = Body::from_stream(ReaderStream::new(file));
    builder
        .body(body)
        .map_err(|err| error_response(StatusCode::INTERNAL_SERVER_ERROR, err.to_string()))
}

pub async fn download(
    Extension(state): Extension<Arc<AppState>>,
    method: Method,
    headers: HeaderMap,
    AxumPath(path): AxumPath<DownloadPath>,
    Query(query): Query<DownloadQuery>,
) -> impl IntoResponse {
    if !is_allowed_type(&path.storage_type) {
        return error_response(StatusCode::BAD_REQUEST, "Invalid type");
    }
    let base_dir = state.main_dir.join(&path.storage_type);
    let Ok(real_path) = safe_path(&base_dir, &path.path) else {
        return error_response(StatusCode::FORBIDDEN, "Access denied");
    };
    if !real_path.is_file() {
        return error_response(StatusCode::NOT_FOUND, "File not found");
    }

    let download_name = build_download_filename(query.filename.as_deref(), &real_path)
        .or_else(|| {
            real_path
                .file_name()
                .map(|value| value.to_string_lossy().to_string())
        })
        .unwrap_or_else(|| "download".to_string());

    if path.storage_type == "archive" && path.path.starts_with("mods/") {
        let Some(mod_id) = path
            .path
            .split('/')
            .nth(1)
            .and_then(|value| value.parse::<u64>().ok())
        else {
            return error_response(StatusCode::NOT_FOUND, "File not found");
        };
        match access_client::resolve_mod_download_access(
            &headers,
            mod_id,
            &state.config.access_service_url,
            state.config.access_service_timeout_seconds,
        )
        .await
        {
            Ok(result) if result.allowed => {}
            Ok(result) => return error_response(StatusCode::FORBIDDEN, result.reason),
            Err(err) => {
                let status = err.status_code.unwrap_or(503);
                return error_response(
                    StatusCode::from_u16(status).unwrap_or(StatusCode::SERVICE_UNAVAILABLE),
                    err.message,
                );
            }
        }
    }

    let mut unpacked_bytes = None;
    if method == Method::HEAD && path.storage_type == "archive" && path.path.starts_with("mods/") {
        let probe_timeout = state
            .config
            .seven_zip_timeout_seconds
            .map(std::time::Duration::from_secs_f64);
        if let Ok((_, is_encrypted, archive_entries)) =
            probe_archive(&real_path, probe_timeout).await
        {
            if !is_encrypted {
                unpacked_bytes = archive_entries_unpacked_bytes(archive_entries.as_deref());
            }
        }
    }

    match open_file_response(
        real_path,
        download_name,
        method == Method::HEAD,
        unpacked_bytes,
    )
    .await
    {
        Ok(response) => response,
        Err(response) => response,
    }
}

#[derive(Default)]
struct UploadParts {
    token: Option<String>,
    storage_type: Option<String>,
    storage_path: Option<String>,
    file_kind: Option<String>,
    file_bytes: Option<Vec<u8>>,
    file_name: Option<String>,
}

async fn read_multipart_parts(mut multipart: Multipart) -> Result<UploadParts, Response> {
    let mut parts = UploadParts::default();
    while let Some(field) = multipart
        .next_field()
        .await
        .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid multipart"))?
    {
        let name = field.name().unwrap_or("").to_string();
        match name.as_str() {
            "file" => {
                parts.file_name = field.file_name().map(|value| value.to_string());
                let bytes = field
                    .bytes()
                    .await
                    .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid file"))?;
                parts.file_bytes = Some(bytes.to_vec());
            }
            "type" => {
                parts.storage_type = Some(
                    field
                        .text()
                        .await
                        .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid type"))?,
                );
            }
            "path" => {
                parts.storage_path = Some(
                    field
                        .text()
                        .await
                        .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid path"))?,
                );
            }
            "file_kind" => {
                parts.file_kind =
                    Some(field.text().await.map_err(|_| {
                        error_response(StatusCode::BAD_REQUEST, "Invalid file kind")
                    })?);
            }
            "token" => {
                parts.token = Some(
                    field
                        .text()
                        .await
                        .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid token"))?,
                );
            }
            _ => {
                let _ = field.bytes().await;
            }
        }
    }
    Ok(parts)
}

pub async fn upload(
    Extension(state): Extension<Arc<AppState>>,
    multipart: Multipart,
) -> impl IntoResponse {
    let parts = match read_multipart_parts(multipart).await {
        Ok(parts) => parts,
        Err(response) => return response,
    };

    let Some(token) = parts.token.as_deref() else {
        return error_response(StatusCode::UNAUTHORIZED, "Token not found");
    };
    if !auth::check_token(&state.config, "upload_file", token) {
        return error_response(StatusCode::FORBIDDEN, "Access denied");
    }

    let storage_type = parts.storage_type.unwrap_or_default();
    if !is_allowed_upload_type(&storage_type) {
        return error_response(StatusCode::BAD_REQUEST, "Invalid type");
    }
    let normalized_file_kind = normalize_file_kind(parts.file_kind.unwrap_or_default(), "");
    if normalized_file_kind.is_empty() {
        return error_response(StatusCode::BAD_REQUEST, "Invalid file kind");
    }
    if storage_type == "avatar" && normalized_file_kind != "img" {
        return error_response(StatusCode::BAD_REQUEST, "Avatar requires image file kind");
    }

    let Some(storage_path) = parts.storage_path else {
        return error_response(StatusCode::BAD_REQUEST, "Invalid path");
    };
    let base_dir = state.main_dir.join(&storage_type);
    let Ok(real_path) = safe_path(&base_dir, &storage_path) else {
        return error_response(StatusCode::FORBIDDEN, "Access denied");
    };
    if let Some(parent) = real_path.parent() {
        let _ = fs::create_dir_all(parent).await;
    }

    let Some(file_bytes) = parts.file_bytes else {
        return error_response(StatusCode::BAD_REQUEST, "Missing file");
    };

    let write_bytes = if normalized_file_kind == "img" {
        if !storage_path.to_lowercase().ends_with(".webp") {
            return error_response(
                StatusCode::BAD_REQUEST,
                "Image storage path must end with .webp",
            );
        }
        match image_bytes_to_webp(&file_bytes, 80) {
            Ok(bytes) => bytes,
            Err(_) => return error_response(StatusCode::BAD_REQUEST, "Image expected"),
        }
    } else {
        file_bytes
    };

    if let Err(err) = fs::write(&real_path, write_bytes).await {
        return error_response(StatusCode::INTERNAL_SERVER_ERROR, err.to_string());
    }

    (StatusCode::CREATED, storage_path).into_response()
}

pub async fn delete(
    Extension(state): Extension<Arc<AppState>>,
    axum::extract::Form(request): axum::extract::Form<DeleteRequest>,
) -> impl IntoResponse {
    let Some(token) = request.token.as_deref() else {
        return error_response(StatusCode::UNAUTHORIZED, "Token not found");
    };
    if !auth::check_token(&state.config, "delete_file", token) {
        return error_response(StatusCode::FORBIDDEN, "Access denied");
    }
    if !is_allowed_type(&request.storage_type) {
        return error_response(StatusCode::BAD_REQUEST, "Invalid type");
    }

    let base_dir = state.main_dir.join(&request.storage_type);
    let Ok(real_path) = safe_path(&base_dir, &request.path) else {
        return error_response(StatusCode::FORBIDDEN, "Access denied");
    };
    if !real_path.is_file() {
        return error_response(StatusCode::NOT_FOUND, "File not found");
    }
    if let Err(err) = fs::remove_file(&real_path).await {
        return error_response(StatusCode::INTERNAL_SERVER_ERROR, err.to_string());
    }

    let mut folder_path = real_path.parent().map(|value| value.to_path_buf());
    let root_dir = base_dir.canonicalize().unwrap_or(base_dir);
    while let Some(current) = folder_path {
        if current == root_dir {
            break;
        }
        match fs::read_dir(&current).await {
            Ok(mut entries) => {
                if entries.next_entry().await.ok().flatten().is_none() {
                    let _ = fs::remove_dir(&current).await;
                    folder_path = current.parent().map(|value| value.to_path_buf());
                    continue;
                }
            }
            Err(_) => {}
        }
        break;
    }

    (StatusCode::OK, "File deleted").into_response()
}
